### PR TITLE
[cmd/opampsupervisor] Add OpAMP supervisor skeleton

### PR DESCRIPTION
**Pull request by evan-bradley**
**Description:**

This is a direct copy of the [supervisor example implementation](https://github.com/open-telemetry/opamp-go/tree/main/internal/examples/supervisor) in the opamp-go repository. I've made some minor changes that are largely centered around linting and making it a little less of an example. Any existing or new TODOs have a linked issue that gives some context.

**Link to tracking Issue:**

Implements https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18461

**Testing:**

No tests right now. I would like for this to have pretty extensive tests, so I will prioritize those as an early follow-up.

**Documentation:**

I've added a README with some basic information and have copied the [design document](https://docs.google.com/document/d/1KtH5atZQUs9Achbce6LiOaJxLbksNJenvgvyKLsJrkc/edit) in Google Docs into a Markdown file that will live next to the Supervisor. Updates can then be made as PRs.

cc @andykellr @atoulme @mhausenblas @portertech @tigrannajaryan

### CODE REVIEW

File: .github/CODEOWNERS
Comment (evan-bradley) [inline .github/CODEOWNERS:25]: If anyone would like to be a code owner, let me know and I will add you.

**Comment by evan-bradley** [comment]
I'd like to wait to add a changelog entry until we actually release a usable binary for the supervisor. I'm open to reconsidering this if anyone feels strongly that we should have an entry.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (tigrannajaryan) [inline cmd/opampsupervisor/supervisor/supervisor.go:205]: The Supervisor also needs to persist the instance id so that it does not change on every Supervisor restart.

See https://docs.google.com/document/d/1KtH5atZQUs9Achbce6LiOaJxLbksNJenvgvyKLsJrkc/edit#heading=h.8szgxmjard0e

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (tigrannajaryan) [inline cmd/opampsupervisor/supervisor/supervisor.go:410]:
```suggestion
// Recalculate the Agent's effective config and if the config changes, signal to the
```

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (tigrannajaryan) [inline cmd/opampsupervisor/supervisor/supervisor.go:633]: https://github.com/open-telemetry/opentelemetry-collector/pull/5402 is already implemented and I think we just need to set `configChanged` to true to trigger the instance id update via config update.

**Review by tigrannajaryan — COMMENTED** [review]
Thanks @evan-bradley 

I just left some comments to link to the relevant parts of the design doc.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (tigrannajaryan) [reply cmd/opampsupervisor/supervisor/commander/commander.go:72]: Please add a TODO and link to the appropriate part of the design document.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (tigrannajaryan) [inline cmd/opampsupervisor/supervisor/supervisor.go:222]: Please add a link to appropriate section in design doc: https://docs.google.com/document/d/1KtH5atZQUs9Achbce6LiOaJxLbksNJenvgvyKLsJrkc/edit#heading=h.8szgxmjard0e

**Review by tigrannajaryan — COMMENTED** [review]
Skeleton LGTM. I left a few small comments to link to the design doc to make implementation of remaining TODOs easier.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go]: Thanks for the link. I've created an issue and linked to it here so we have something to track the work.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/supervisor.go:165]: Done.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/commander/commander.go:73]: Done.

**Comment by evan-bradley** [comment]
@tigrannajaryan Thank you for the review and for the links to the design document. I've [written issues](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues?q=is%3Aissue+is%3Aopen+%5Bcmd%2Fopampsupervisor%5D) and linked them by the TODOs so we can track the work for each and discuss any open points if necessary.

**Review by tigrannajaryan — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@andykellr if you have time it would be great if you can review this.

**Comment by tigrannajaryan** [comment]
@open-telemetry/collector-contrib-approvers please review.

File: .github/CODEOWNERS
Comment (TylerHelmuth) [inline .github/CODEOWNERS:25]: Are any of these teams appropriate code owners? https://github.com/orgs/open-telemetry/teams?query=opamp

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (djaglowski) [inline cmd/opampsupervisor/supervisor/commander/commander.go:156]: I may be missing something, but how is this channel associated with the agent process?

Comment (tigrannajaryan) [reply cmd/opampsupervisor/supervisor/commander/commander.go:156]: I may have written this code (guilty). Looking at it again, `finished` is possibly redundant and can be replaced by `c.waitCh`.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (cpheps) [inline cmd/opampsupervisor/supervisor/commander/commander.go:168]: Could this just be collapsed to a single context. Context will be done if the parent or the timeout occurs. In either case we break it looks like.

```go
waitCtx := context.WithTimout(ctx, 10 * time.Second)

select {
 case <-waitCtx.Done():
    break
...
}
```

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (cpheps) [inline cmd/opampsupervisor/supervisor/commander/commander.go:91]: Will this work if `Stop` is called then `Restart` is called? I'm not sure what `c.cmd.Process.Signal` does if it's called multiple times in a row.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/commander/commander.go]: Good catch. I've eliminated the extra channel.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/commander/commander.go]: Updated to only use `doneCh`.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/commander/commander.go]: Good catch. I've updated this to use a context.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/commander/commander.go:96]: If `Stop` is called, then `Restart` is called, the subsequent call to `Stop` is a no-op since there's a check to see whether there is a running process to stop.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/commander/commander.go:180]: I've managed to eliminate the `select` statement, so we're handling this in that case now.

File: cmd/opampsupervisor/supervisor/commander/commander.go
Comment (cpheps) [reply cmd/opampsupervisor/supervisor/commander/commander.go:66]: Looks good to me

**Review by cpheps — APPROVED** [review]

**Comment by evan-bradley** [comment]
@atoulme This should be ready to go, could you please review as the sponsor for this component?

**Review by mwear — APPROVED** [review]

File: cmd/opampsupervisor/supervisor/healthchecker/healthchecker.go
Comment (atoulme) [inline cmd/opampsupervisor/supervisor/healthchecker/healthchecker.go:24]: This is probably a bit too immature and we should look at using confighttp.HTTPClientSettings to set this up, in a later PR

File: cmd/opampsupervisor/supervisor/opamplogger.go
Comment (atoulme) [inline cmd/opampsupervisor/supervisor/opamplogger.go:14]: I don't think it does that. It adds a couple convenience methods that allows to compose strings using Sprintf. I am not fond of this, as zap fields are meant to reduce the cost of concatenation in case the log doesn't happen, but it ends up being a matter of personal preference. Not blocking this PR.

File: cmd/opampsupervisor/supervisor/supervisor.go
Comment (atoulme) [inline cmd/opampsupervisor/supervisor/supervisor.go:238]: does this panic?

**Review by atoulme — APPROVED** [review]
LGTM

File: cmd/opampsupervisor/supervisor/healthchecker/healthchecker.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/healthchecker/healthchecker.go:26]: That's a fair point. I've opened https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/23344 to address this and linked to it in this file.

File: cmd/opampsupervisor/supervisor/opamplogger.go
Comment (evan-bradley) [reply cmd/opampsupervisor/supervisor/opamplogger.go]: I agree it's not an ideal solution, it was just a way to deal with the fact that Zap wants `zap.Field` and the OpAMP client logging interface accepts `any`.

You made me realize I completely forgot about `zap.SugaredLogger`, so I've switched to just passing the OpAMP client the sugared logger directly, which obviates the need for this adapter struct.

**Review by codeboten — APPROVED** [review]

**Pull request merged**
